### PR TITLE
Jump to entry key cli (WIP)

### LINKED ISSUE #333: JabRef Groups interface should work at low resolutions 1366x768
1366x768 - top and bottom bar is there, "OK" button cannot be pressed

### PR DESCRIPTION
**Pull request by u7282852**
Addresses Issue: https://github.com/koppor/jabref/issues/540
Command line now supports option "jumpToEntryKey" or "-j" folowed by one entrykey argument. 
Upon start up, jabref will search the imported files for the entrykey and focus this library and select this entry. 

However, I am experiencing issues when there are previously opened databases that are not opened through the command line. The last database within the previously opened databases seems to always be the library focused on start up as they seem to run on a different thread or something. 

Currently the key is searched only in the imported files specified in the commandline [BIBTEX_FILE] fields. I also request some assistance as to how I can search the previously opened/edited databases as well for the entrykey. 

### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/cli/ArgumentProcessor.java [hunk 1]
```diff
@@ -773,6 +777,18 @@ private Optional<ParserResult> fetch(String fetchCommand) {
...
+    private void jumpToEntryKey(List<ParserResult> loaded, String citationKey) {
+        // search for this key in imported files
+        for (ParserResult parserResult : loaded) {
+            Optional<BibEntry> entry = parserResult.getDatabase().getEntryByCitationKey(citationKey);
+            if (entry.isPresent()) {
+                parserResult.setEntryToFocus(entry.get());
+                return;
+            }
+        }
+        LOGGER.error("Could not find given citation key to jump to");
```
Comment (koppor) [inline src/main/java/org/jabref/cli/ArgumentProcessor.java:789]:
```suggestion
        LOGGER.error("Could not find citation key {} to jump to", citationKey);
```

Here, `System.out.println` is OK, because it was a CLI which outputs to the console. Please also add the output to the console. -- Maybe, even replace logger completely:

    System.out.println("Could not find citation key %0 in any library", citationKey);

See https://devdocs.jabref.org/code-howtos/localization.html for steps for the localiaztion.

File: src/main/java/org/jabref/cli/ArgumentProcessor.java [hunk 2]
```diff
@@ -773,6 +777,18 @@ private Optional<ParserResult> fetch(String fetchCommand) {
         }
     }
 
+    private void jumpToEntryKey(List<ParserResult> loaded, String citationKey) {
+        // search for this key in imported files
```
Comment (koppor) [inline src/main/java/org/jabref/cli/ArgumentProcessor.java:781]: Remove this line, because the for each loop makes it clear somehow.

File: src/main/java/org/jabref/cli/JabRefCLI.java [hunk 3]
```diff
@@ -288,6 +296,14 @@ private static Options getOptions() {
                 .argName("CITEKEY1[,CITEKEY2][,CITEKEYn] | PDF1[,PDF2][,PDFn] | all")
                 .build());
 
+        options.addOption(Option
+                .builder("j")
+                .longOpt("jumpToEntryKey")
```
Comment (koppor) [inline src/main/java/org/jabref/cli/JabRefCLI.java:301]: Rename to `jumpToKey`. (We use `CitationKey` throughout JabRef. Sometimes also `BibtexKey`. Never `EntryKey`. With `key` you are on the safe side ^^)

File: src/main/java/org/jabref/cli/JabRefCLI.java [hunk 4]
```diff
@@ -288,6 +296,14 @@ private static Options getOptions() {
...
+                .desc(String.format("%s: '%s'", Localization.lang("Jump to the BibEntry of the given key."), "-j key"))
```
Comment (koppor) [inline src/main/java/org/jabref/cli/JabRefCLI.java:302]:
```suggestion
                .desc(String.format("%s: '%s'", Localization.lang("Jump to the entry of the given citation key."), "-j key"))
```

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 5]
```diff
@@ -201,14 +202,29 @@ private void openDatabases() {
                                         .orElse(preferencesService.getGuiPreferences()
                                                                  .getLastFocusedFile())
                                         .toAbsolutePath();
+        // check whether there is a jump to entry and need to focus different file
+        boolean focusDifferent = false;
+        Optional<BibEntry> focusedEntry = Optional.empty();
+        for (ParserResult parserResult : parserResults) {
+            // Make sure this parser result is its own library instead of imported BibTex entries
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:209]: Write "BibTeX" always with capital `X`

File: src/main/java/org/jabref/gui/JabRefGUI.java [hunk 6]
```diff
@@ -201,14 +202,29 @@ private void openDatabases() {
...
 
         // Add all bibDatabases databases to the frame:
         boolean first = false;
         for (ParserResult parserResult : parserResults) {
             // Define focused tab
-            if (parserResult.getPath().filter(path -> path.toAbsolutePath().equals(focusedFile)).isPresent()) {
+            if (parserResult.getPath().filter(path -> path.toAbsolutePath().equals(focusedFile)).isPresent() && !focusDifferent) {
                 first = true;
             }
+            if (parserResult.getEntryToFocus().isPresent() && focusDifferent) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/JabRefGUI.java:224]: Should that be `} else if (...`?

**Review by koppor — CHANGES_REQUESTED** [review]
Some small comments.

Please include it and submit to the main repo

**Comment by koppor** [comment]
> I also request some assistance as to how I can search the previously opened/edited databases as well for the entrykey.

I found `org.jabref.gui.JabRefGUI#openLastEditedDatabases`, which openes the libraries. Maybe this helps to understand the control flow of opening

File: CHANGELOG.md [hunk 7]
```diff
@@ -20,6 +20,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We added a fetcher for [LOBID](https://lobid.org/resources/api) resources. [koppor#386](https://github.com/koppor/jabref/issues/386)
 - When in `biblatex` mode, the [integrity check](https://docs.jabref.org/finding-sorting-and-cleaning-entries/checkintegrity) for journal titles now also checks the field `journal`.
 - We added support for pushing citations to [TeXShop](https://pages.uoregon.edu/koch/texshop/) on macOS [forum#2699](https://discourse.jabref.org/t/push-to-texshop-mac/2699).
+- We added ability to jump to an entry in the command line using -j CITATIONKEY [BIBTEXFILES]. [koppor#540](https://github.com/koppor/jabref/issues/540)
```
Comment (koppor) [inline CHANGELOG.md:23]:
```suggestion
- We added ability to jump to an entry in the command line using `-j CITATIONKEY [BIBTEXFILES]`. [koppor#540](https://github.com/koppor/jabref/issues/540)
```

File: src/main/java/org/jabref/cli/JabRefCLI.java [hunk 8]
```diff
@@ -298,10 +298,10 @@ private static Options getOptions() {
 
         options.addOption(Option
                 .builder("j")
-                .longOpt("jumpToEntryKey")
-                .desc(String.format("%s: '%s'", Localization.lang("Jump to the BibEntry of the given key."), "-j key"))
+                .longOpt("jumpToKey")
+                .desc(String.format("%s: '%s'", Localization.lang("Jump to the entry of the given citation key."), "-j key"))
                 .hasArg()
-                .argName("ENTRYKEY")
+                .argName("CITEKEY")
```
Comment (koppor) [inline src/main/java/org/jabref/cli/JabRefCLI.java:304]:
```suggestion
                .argName("CITATIONKEY")
```

File: src/main/java/org/jabref/gui/LibraryTab.java [hunk 9]
```diff
@@ -222,7 +222,7 @@ public Node createLoadingAnimationLayout() {
     public void onDatabaseLoadingStarted() {
         Node loadingLayout = createLoadingAnimationLayout();
         getMainTable().placeholderProperty().setValue(loadingLayout);
-        frame.addTab(this, true);
+        frame.addTab(this, false);
```
Comment (koppor) [inline src/main/java/org/jabref/gui/LibraryTab.java:225]: Why was this changed? --> Add a java comment why `false` is passed.

**Review by koppor — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/LibraryTab.java (hunk 9 above)
Comment (u7282852) [reply src/main/java/org/jabref/gui/LibraryTab.java:227]: I noticed that either the first database given as argument, or the last loaded previously opened file will be focused later regardless. This change also stopped previously opened libraries overriding the focus of a library that was already focused given the openFies method in OpenDatabaseAction ran on a different thread.

**Comment by koppor** [comment]
Follow-up at https://github.com/JabRef/jabref/pull/10578. Closing this one. The other PR is the same branch, but more committers can see it!
